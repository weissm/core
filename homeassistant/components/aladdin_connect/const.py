"""Constants for the Aladdin Connect Genie integration."""

DOMAIN = "aladdin_connect"

OAUTH2_AUTHORIZE = "https://app.aladdinconnect.net/login.html"
OAUTH2_TOKEN = "https://twdvzuefzh.execute-api.us-east-2.amazonaws.com/v1/oauth2/token"
