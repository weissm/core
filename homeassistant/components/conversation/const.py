"""Const for conversation integration."""

DOMAIN = "conversation"
DEFAULT_EXPOSED_ATTRIBUTES = {"device_class"}
HOME_ASSISTANT_AGENT = "conversation.home_assistant"
OLD_HOME_ASSISTANT_AGENT = "homeassistant"

ATTR_TEXT = "text"
ATTR_LANGUAGE = "language"
ATTR_AGENT_ID = "agent_id"
ATTR_CONVERSATION_ID = "conversation_id"

SERVICE_PROCESS = "process"
SERVICE_RELOAD = "reload"
