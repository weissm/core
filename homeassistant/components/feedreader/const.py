"""Constants for RSS/Atom feeds."""

from datetime import timedelta

DOMAIN = "feedreader"

CONF_MAX_ENTRIES = "max_entries"
DEFAULT_MAX_ENTRIES = 20
DEFAULT_SCAN_INTERVAL = timedelta(hours=1)
