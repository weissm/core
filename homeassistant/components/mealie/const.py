"""Constants for the Mealie integration."""

import logging

DOMAIN = "mealie"

LOGGER = logging.getLogger(__package__)
