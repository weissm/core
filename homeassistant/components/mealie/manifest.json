{
  "domain": "mealie",
  "name": "Mealie",
  "codeowners": ["@joostlek"],
  "config_flow": true,
  "documentation": "https://www.home-assistant.io/integrations/mealie",
  "integration_type": "service",
  "iot_class": "local_polling",
  "requirements": ["aiomealie==0.4.0"]
}
