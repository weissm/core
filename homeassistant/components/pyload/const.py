"""Constants for the pyLoad integration."""

DOMAIN = "pyload"

DEFAULT_HOST = "localhost"
DEFAULT_NAME = "pyLoad"
DEFAULT_PORT = 8000
