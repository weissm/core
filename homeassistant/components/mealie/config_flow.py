"""Config flow for Mealie."""

from typing import Any

from aiomealie import MealieAuthenticationError, MealieClient, MealieConnectionError
import voluptuous as vol

from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
from homeassistant.const import CONF_API_TOKEN, CONF_HOST
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DOMAIN, LOGGER

SCHEMA = vol.Schema(
    {
        vol.Required(CONF_HOST): str,
        vol.Required(CONF_API_TOKEN): str,
    }
)


class MealieConfigFlow(ConfigFlow, domain=DOMAIN):
    """Mealie config flow."""

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle a flow initialized by the user."""
        errors: dict[str, str] = {}
        if user_input:
            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
            client = MealieClient(
                user_input[CONF_HOST],
                token=user_input[CONF_API_TOKEN],
                session=async_get_clientsession(self.hass),
            )
            try:
                await client.get_mealplan_today()
            except MealieConnectionError:
                errors["base"] = "cannot_connect"
            except MealieAuthenticationError:
                errors["base"] = "invalid_auth"
            except Exception:  # noqa: BLE001
                LOGGER.exception("Unexpected error")
                errors["base"] = "unknown"
            else:
                return self.async_create_entry(
                    title="Mealie",
                    data=user_input,
                )
        return self.async_show_form(
            step_id="user",
            data_schema=SCHEMA,
            errors=errors,
        )
