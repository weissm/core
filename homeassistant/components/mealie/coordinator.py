"""Define an object to manage fetching Mealie data."""

from __future__ import annotations

from datetime import timedelta
from typing import TYPE_CHECKING

from aiomealie import (
    MealieAuthenticationError,
    MealieClient,
    MealieConnectionError,
    Mealplan,
    MealplanEntryType,
)

from homeassistant.const import CONF_API_TOKEN, CONF_HOST
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryError
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
import homeassistant.util.dt as dt_util

from .const import LOGGER

if TYPE_CHECKING:
    from . import MealieConfigEntry

WEEK = timedelta(days=7)


class MealieCoordinator(DataUpdateCoordinator[dict[MealplanEntryType, list[Mealplan]]]):
    """Class to manage fetching Mealie data."""

    config_entry: MealieConfigEntry

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass, logger=LOGGER, name="Mealie", update_interval=timedelta(hours=1)
        )
        self.client = MealieClient(
            self.config_entry.data[CONF_HOST],
            token=self.config_entry.data[CONF_API_TOKEN],
            session=async_get_clientsession(hass),
        )

    async def _async_update_data(self) -> dict[MealplanEntryType, list[Mealplan]]:
        next_week = dt_util.now() + WEEK
        try:
            data = (
                await self.client.get_mealplans(dt_util.now().date(), next_week.date())
            ).items
        except MealieAuthenticationError as error:
            raise ConfigEntryError("Authentication failed") from error
        except MealieConnectionError as error:
            raise UpdateFailed(error) from error
        res: dict[MealplanEntryType, list[Mealplan]] = {
            MealplanEntryType.BREAKFAST: [],
            MealplanEntryType.LUNCH: [],
            MealplanEntryType.DINNER: [],
            MealplanEntryType.SIDE: [],
        }
        for meal in data:
            res[meal.entry_type].append(meal)
        return res
