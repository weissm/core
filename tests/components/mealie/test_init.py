"""Tests for the Mealie integration."""

from unittest.mock import AsyncMock

from aiomealie import MealieAuthenticationError, MealieConnectionError
import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.mealie.const import DOMAIN
from homeassistant.config_entries import ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr

from . import setup_integration

from tests.common import MockConfigEntry


async def test_device_info(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_mealie_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
    device_registry: dr.DeviceRegistry,
) -> None:
    """Test device registry integration."""
    await setup_integration(hass, mock_config_entry)
    device_entry = device_registry.async_get_device(
        identifiers={(DOMAIN, mock_config_entry.entry_id)}
    )
    assert device_entry is not None
    assert device_entry == snapshot


async def test_load_unload_entry(
    hass: HomeAssistant,
    mock_mealie_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test load and unload entry."""
    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.LOADED

    await hass.config_entries.async_remove(mock_config_entry.entry_id)
    await hass.async_block_till_done()

    assert mock_config_entry.state is ConfigEntryState.NOT_LOADED


@pytest.mark.parametrize(
    ("exc", "state"),
    [
        (MealieConnectionError, ConfigEntryState.SETUP_RETRY),
        (MealieAuthenticationError, ConfigEntryState.SETUP_ERROR),
    ],
)
async def test_initialization_failure(
    hass: HomeAssistant,
    mock_mealie_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
    exc: Exception,
    state: ConfigEntryState,
) -> None:
    """Test initialization failure."""
    mock_mealie_client.get_mealplans.side_effect = exc

    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is state
