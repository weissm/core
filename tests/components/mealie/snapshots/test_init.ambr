# serializer version: 1
# name: test_device_info
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': <DeviceEntryType.SERVICE: 'service'>,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'mealie',
        '01J0BC4QM2YBRP6H5G933CETT7',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'name': 'Mealie',
    'name_by_user': None,
    'serial_number': None,
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
